docs/rtcalc.1
.\" Automatically generated by Pandoc 3.10
.\"
.TH "rtcalc" "1" "June 2026" "rtcalc release 1.28.8"
.SH NAME
rtcalc \- Interactive CLI math program
.SH SYNOPSIS
\f[CR]rtcalc\f[R] [flags]
.SH DESCRIPTION
Takes in user input interactively and in real time, calculating the
formula provided by the user and outputting it.
.SH OPTIONS
.TP
\f[CR]help\f[R]
Show a help menu, providing basic usage info.
.TP
\f[CR]prompt=[str]\f[R]
Replace default shell prompt with user provided one.
Note that the space beetwen input and prompt is NOT automatically
provided.
.TP
\f[CR]syntax\-highlighting\f[R]
Enable input syntax highlighting, coloring different components of the
input.
.TP
\f[CR]precision=[num]\f[R]
Sets the printed result\(cqs floating point precision to a user set
value.
.TP
\f[CR]define:[name]=[val]\f[R]
Defines a new variable of name with a value ([value]).
The value field is calculated in the same way that a formula would
interactively, meaning that even functions apply.
And naturally, you may also use variables defined before in the formula.
.SH BASIC USAGE
After starting the program, type in the formula to be calculated,
following basic math principles.
The program will automatically error out on invalid formulas or syntax.
.SH SYNTAX AND EXAMPLES
Type in two operands and an operator.
.PP
\f[CR]x + y\f[R]
.PP
The program supports more than one operand, as long as there are enough
operators.
.PP
\f[CR]x + y \- z\f[R]
.PP
Each operator has an order of operations, that being:
.PP
\f[CR]+\f[R], \f[CR]\-\f[R]: Third priority.
.PP
\f[CR]*\f[R], \f[CR]/\f[R], \f[CR]%\f[R]: Second priority.
.PP
\f[CR]\(ha\f[R]: First priority.
.PP
You may introduce parentheses.
.PP
\f[CR]x * y \(ha (z + z)\f[R]
.PP
Parentheses may be nested.
.PP
\f[CR]x + (x + (y * (z + x)))\f[R]
.PP
Functions are implemented.
Their results are treated as operands.
.PP
\f[CR]sqrt[x] + y\f[R]
.PP
You may write formulas inside of functions.
.PP
\f[CR]sqrt[x * y]\f[R]
.PP
Defined variables may be integrated seamlessly, being treated as regular
operands.
.PP
\f[CR]{x} * {y}\f[R]
.PP
Specific formulas will required two arguments.
Separate each with a semicolon.
.PP
\f[CR]logx[x ; y]\f[R]
.SH OPERATORS
.TP
\f[CR]x + y\f[R]
Perform addition.
.TP
\f[CR]x \- y\f[R]
Perform subtraction.
.TP
\f[CR]x * y\f[R]
Perform addition.
.TP
\f[CR]x / y\f[R]
Perform addition.
.TP
\f[CR]x % y\f[R]
Perform division, and return the rest.
.TP
\f[CR]x \(ha y\f[R]
Perform exponentiation.
.SH FUNCTIONS
.SS Trigonometric
Note that all trigonometric functions return in radians.
.TP
\f[CR]sin[x]\f[R]
Perform the sine of `x'.
.TP
\f[CR]cos[x]\f[R]
Perform the cosine of `x'.
.TP
\f[CR]tan[x]\f[R]
Perform the tangent of `x'.
.TP
\f[CR]sinh[x]\f[R]
Perform the hyperbolic sine of `x'.
.TP
\f[CR]cosh[x]\f[R]
Perform the hyperbolic cosine of `x'.
.TP
\f[CR]tan[x]\f[R]
Perform the hyperbolic tangent of `x'.
.TP
\f[CR]asin[x]\f[R]
Perform the reverse sine of `x'.
.TP
\f[CR]acos[x]\f[R]
Perform the reverse cosine of `x'.
.TP
\f[CR]atan[x]\f[R]
Perform the reverse tangent of `x'.
.TP
\f[CR]asinh[x]\f[R]
Perform the reverse hyperbolic sine of `x'.
.TP
\f[CR]acosh[x]\f[R]
Perform the reverse hyperbolic cosine of `x'.
.TP
\f[CR]atanh[x]\f[R]
Perform the reverse hyperbolic tangent of `x'.
.TP
\f[CR]atan2[x]\f[R]
Get the angle theta from the conversion of rectangular coordinates (x,
y) to polar coordinates (r, theta).
.TP
\f[CR]csc[x]\f[R]
Perform the cosecant of `x'.
.TP
\f[CR]sec[x]\f[R]
Perform the secant of `x'.
.TP
\f[CR]cot[x]\f[R]
Perform the cotangent of `x'.
.TP
\f[CR]csch[x]\f[R]
Perform the hyperbolic cosecant of `x'.
.TP
\f[CR]sech[x]\f[R]
Perform the hyperbolic secant of `x'.
.TP
\f[CR]coth[x]\f[R]
Perform the hyperbolic cotangent of `x'.
.SS Conversion
.TP
\f[CR]deg[x]\f[R]
Convert radians to degrees.
.TP
\f[CR]rad[x]\f[R]
Convert degrees to radians.
.SS Rounding
.TP
\f[CR]floor[x]\f[R]
Get the floor of `x'.
.TP
\f[CR]ceil[x]\f[R]
Get the ceiling of `x'.
.TP
\f[CR]trunc[x]\f[R]
Get the integer part of `x'.
.TP
\f[CR]round[x]\f[R]
Round `x' to the nearest integer.
.TP
\f[CR]rndflr[x, y]\f[R]
Round `x\(cqs floor to \(cqy'.
.TP
\f[CR]rndcei[x, y]\f[R]
Round `x\(cqs ceiling to \(cqy'.
.SS Exponential
.TP
\f[CR]sqrt[x]\f[R]
Perform the square root of `x'.
.TP
\f[CR]cbrt[x]\f[R]
Perform the cube root of `x'.
.TP
\f[CR]log[x]\f[R]
Perform the natural logarithm of `x'.
.TP
\f[CR]log10[x]\f[R]
Perform the base 10 logarithm of `x'.
.TP
\f[CR]log2[x]\f[R]
Perform the base 2 logarithm of `x'.
.TP
\f[CR]logx[x, y]\f[R] (aka \f[CR]log_x()\f[R])
Perform the base `y' logarithm of `x'.
.TP
\f[CR]exp[x]\f[R]
Perform the value of E \(ha `x'.
.TP
\f[CR]exp2[x]\f[R]
Perform the value of 2 \(ha `x'.
.TP
\f[CR]pow[x, y]\f[R]
Get the result of `x' to the power of `y'.
Equivalent to operator `\f[CR]\(ha\f[R]'.
.SS Geometry
.TP
\f[CR]hypot[x, y]\f[R]
Get the length of the length of the hypotenuse of a right angle triangle
with sides of length `x' and `y', from origin (0, 0).
.SS Special Functions
.TP
\f[CR]gamma[x]\f[R]
Perform the gamma of `x'.
.TP
\f[CR]lgamma[x]\f[R]
Perform the natural logarithm of the absolute gamma of `x'.
.TP
\f[CR]erf[x]\f[R]
Perform the value of the error function at `x'.
.TP
\f[CR]erfc[x]\f[R]
Perform the value of the complementary error function at `x'.
.SS Miscellaneous
.TP
\f[CR]abs[x]\f[R]
Get the absolute of `x'.
.TP
\f[CR]fmax[x, y]\f[R]
Get the highest value of a floating `x' and `y'.
.TP
\f[CR]fmin[x, y]\f[R]
Get the lowest value of a floating `x' and `y'.
.SH NAVIGATION
The program offers basic shell\-like navigation options.
.TP
\f[CR]CTRL + A\f[R]
Move cursor to start of input buffer.
.TP
\f[CR]CTRL + E\f[R]
Move cursor to end of input buffer.
.TP
\f[CR]CTRL + K\f[R]
Clear everything right of cursor.
.TP
\f[CR]CTRL + X\f[R]
Clear input buffer.
.TP
\f[CR]CTRL + W\f[R]
Delete token by token.
.TP
\f[CR]RIGHT ARROW\f[R]
Move cursor to the right.
.TP
\f[CR]LEFT ARROW\f[R]
Move cursor to the left.
.TP
\f[CR]CTRL + RIGHT ARROW\f[R]
Move cursor to the right.
.TP
\f[CR]CTRL + LEFT ARROW\f[R]
Move cursor to the left.
.SH SPECIFIC DETAILS
The input buffer has a limit of 2048 characters.
.PP
The result buffer has a limit of 1024 characters.
.SH DEVELOPER NOTES
I hope that you like this program <3
.SH AUTHORS
undeleted.


src/rtcalc.c
#include <asm-generic/errno-base.h>
#include <asm-generic/ioctls.h>
#include <ctype.h>
#include <signal.h>
#include <stdint.h>
#include <termios.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include "functions.h"
#include "definitions.h"
#include "utilities.h"

struct termios backup = { 0 };
int retCode = 0;
volatile sig_atomic_t needsRedraw = 0;

int main (int argc, char *argv[]) {
    if (!isatty(STDIN_FILENO)) {
        fprintf(stderr, "Not being operated in a terminal - terminating.\n");
        return USER_MISTAKE;
    }

    unsigned long long precision = 6;
    char *prompt = ">>> ";
    // see definitions for the flags
    unsigned char globalFlags = 0;

    // environment variables
    // maybe only allocate if needed in the future with <alloca.h>?
    struct variable variables[MAX_VARIABLES + 1] = { 0 };
    unsigned short varCount = 0; // how many slots were used up + index

    // loop trough stuffies and match argument
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "help")) {
            helpMenu(NULL, 0);

        } else if (!strncmp(argv[i], "prompt", 6)) {
            // set custom prompt
            char *start = strchr(argv[i], '=');
            if (!start || !*(start + 1))
                helpMenu("\n\e[31mError: Prompt not defined properly.\e[0,\n", USER_MISTAKE);
            start++;
            prompt = start;

        } else if (!strcmp(argv[i], "syntax-highlighting")) {
            // define syntax highlightning
            globalFlags |= USE_PRETTY_COLORS;

        } else if (!strncmp(argv[i], "precision", 9)) {
            // define exact precision
            char *start = strchr(argv[i], '=');
            if (!start || !*(start + 1))
                helpMenu("\n" ERROR_CLR "\e[31mError: Precision not defined properly.\e[0,\n", USER_MISTAKE);

            char *check = ++start;
            precision = strtoul(start, &check, 10);

            if (check == start || *check != '\0')
                helpMenu("\n" ERROR_CLR "Error: Precision is an invalid number.\e[0,\n", USER_MISTAKE);

        } else if (!strncmp(argv[i], "define", 5)) {
            // define env variables
            // if there are already too many
            if ((varCount + 1) > MAX_VARIABLES)
                helpMenu("\n" ERROR_CLR "\e[31mError: Too many variables defined.\e[0,\n", USER_MISTAKE);

            char *name = strchr(argv[i], ':');
            if (!name)
                helpMenu("\n" ERROR_CLR "\e[31mError: Definition used incorrectly.\e[0,\n", USER_MISTAKE);
            name++; // name is at start of name
            if (!*name)
                helpMenu("\n" ERROR_CLR "\e[31mError: Definition name was not set.\e[0,\n", USER_MISTAKE);

            char *valuePtr = strchr(argv[i], '=');
            if (!valuePtr)
                helpMenu("\n" ERROR_CLR "\e[31mError: Definition has no value.\e[0,\n", USER_MISTAKE);
            char *eq = valuePtr;

            if (valuePtr == (name))
                helpMenu("\n" ERROR_CLR "\e[31mError: Definition name is invalid.\e[0,\n", USER_MISTAKE);

            if (*(valuePtr + 1) == '\0')
                helpMenu("\n" ERROR_CLR "\e[31mError: Definition value is invalid.\e[0,\n", USER_MISTAKE);

            valuePtr++;

            // normalize comma decimal to dot
            for (char *p = valuePtr; *p; p++) if (*p == ',') *p = '.';

            // run calculation stack
            int childHighestPrio = 0;
            char ret = 0;
            if ((ret = validateBuffer(valuePtr, &childHighestPrio, variables)))
                helpMenu("\n" ERROR_CLR "\e[31mError: Variable couldn't be expanded due to an invalid formula.\e[0,\n", USER_MISTAKE);

            // define
            variables[varCount].value = calculateBuffer(valuePtr, childHighestPrio, variables);
            variables[varCount].name  = name;
            // set null terminator, allowed under C99 section 5.1.2.2.1 :tongue:
            *eq = '\0';
            variables[varCount].len   = strlen(name);

            varCount++;

        } else {
            helpMenu("\n\e[31mError: Improper flag used.\e[0m\n", USER_MISTAKE);
        }
    }
    const size_t promptLen = strlen(prompt);

    // handle signals
    signal(SIGINT,   handleCtrlC);
    signal(SIGWINCH, handleTermResize);

    struct termios newMode = { 0 };
    // populate backup and define new mode
    if (tcgetattr(STDIN_FILENO, &backup) == -1) {
        fprintf(stderr, "Failed to backup term attributes: %s\n", strerror(errno));
        return CODE_MISTAKE;
    }
    atexit(restoreTerminal); // set up putting the terminal back to normal

    // change terminal mode to non-canonical for magic
    newMode = backup;
    newMode.c_lflag &= ~(ICANON | ECHO); // don't wait for new line and dont say anything
    newMode.c_cc[VMIN]  = 1; // get one char at a time
    newMode.c_cc[VTIME] = 0;
    if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &newMode) == -1) {
        fprintf(stderr, "Failed to change terminal mode: %s\n", strerror(errno));
        return CODE_MISTAKE;
    }
    putchar('\n');

    // get the absolute row
    printf("\e[6n");
    char getRowBuf[16] = { 0 };
    uint32_t startRow = 0; // holds the absolute row of program start
    scanf("\e[%d;%*dR", &startRow);

    // main loop to pupulate main buffer
    char calcBuffer[BUFFER_SIZE] = { 0 };
    ssize_t cursorPos = 0;

    // show prompt
    printf("\n%s", prompt);
    char result[RESULT_SIZE] = { 0 };
    size_t resSize = 0;

    // term info
    struct winsize ws = { 0 };
    if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) != 0) {
        fprintf(stderr, "Failed to get terminal info: %s\n", strerror(errno));
        return CODE_MISTAKE;
    }

    while(1) {
        int highestPrio = 0;
        uint8_t ret = validateBuffer(calcBuffer, &highestPrio, variables); // find errors
        if (cursorPos >= 4096) ret = E_INPUT_SIZE_LIMIT; // input size limit

        // define result
        if (!ret) {
            if (cursorPos == 0 && calcBuffer[0] == '\0') {
                memset(result, '\0', sizeof(result));
                memcpy(result, WELCOME, sizeof(result));

            } else {

                defaultPrecision calc = calculateBuffer(calcBuffer, highestPrio, variables);
                if ((resSize = snprintf(NULL, 0, FORMAT_WITH_PRECISION, (int)precision, calc)) >= 2048) ret = 12;

                if (ret) {
                    snprintf(result, sizeof(result), "Can't calculate: %s", retToStr(ret));

                } else {
                    snprintf(result, sizeof(result), FORMAT_WITH_PRECISION, (int)precision, calc);
                }

            }
        } else {
            snprintf(result, sizeof(result), "Can't calculate: %s", retToStr(ret));
        }

        printf("\e[%d;1H\e[J"           // return cursor back to start
               "\e[A\r\e[2K%s\r\e[B%s", // replace result and prompt
               startRow, result, prompt
        );

        // print buffer itself
        if (globalFlags & USE_PRETTY_COLORS) {
            // yes colors >:3
            if (!ret) printBufColored(calcBuffer);
            else printf("%s%s\e[0m", ERROR_CLR, calcBuffer);

        } else {
            // no colors :(
            printf("%s%s\e[0m", ret ? ERROR_CLR : "", calcBuffer);
        }

        // we don't the user to type with a big ass buffer
        if (ret >= E_INPUT_SIZE_LIMIT) continue;

        ssize_t len = strlen(calcBuffer);

        // if sigwinch was detected
        if (needsRedraw) {
            needsRedraw = 0;
            if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) != 0) {
                fprintf(stderr, "Failed to get terminal info: %s\n", strerror(errno));
                return CODE_MISTAKE;
            }
            // move everything back to start and clear
            printf("\e[%d;1H\e[J", startRow);
            continue;
        }

        // move cursor, wrapping on terminal edge
        ssize_t remaining = labs(cursorPos - len);
        ssize_t currentCol = (promptLen + 1 + len) % ws.ws_col;
        if (!(currentCol - 1)) putchar('\n');
        // actual movement
        while (remaining > 0) {
            printf("\e[D");
            remaining--;
            currentCol--;

            // if we hit the edge
            if (currentCol == 1 && remaining > 0) {
                printf("\e[A"   // move on line up
                    "\e[999C"); // move all the way to the right
                currentCol = ws.ws_col;
            }
        }

        char c = getchar();
        // prevent newlines
        if (c == '\n' || c == 0x0C) continue;

        // escape character
        if (c == 0x1B) {
            c = getchar();
            if (c == '[') {
                c = getchar();

                switch (c) {
                    // movement keys
                    case 'C': {
                        // right arrow
                        if (cursorPos < len) cursorPos++;
                        continue;
                        break;
                    }
                    case 'D': {
                        // left arrow
                        if (cursorPos > 0) cursorPos--;
                        continue;
                        break;
                    }

                    case '1': {
                        char seq[8] = {0};
                        int i = 0;
                        while ((seq[i] = getchar()) != EOF && i < 7) {
                            if (seq[i] >= 'A' && seq[i] <= '~') break; // final char
                            i++;
                            c = seq[i];
                        }

                        if (!strcmp(seq, ";5D")) {
                            // control+left, aka move left (word)
                            uint8_t leftWhite = 0;

                            if (cursorPos > 0) {
                                if (strchr(DELIMITERS, calcBuffer[cursorPos - 1])) {
                                    // if the cursor is already at a delimiter,
                                    // move only delimiters on loop
                                    while (cursorPos > 0 && strchr(DELIMITERS, calcBuffer[cursorPos - 1])) {
                                        // move on loop
                                        cursorPos--;
                                    }

                                } else {
                                    while (cursorPos > 0) {
                                        // skip trailling whitespace
                                        if (!leftWhite) {
                                            if (!isspace(calcBuffer[cursorPos - 1])) leftWhite = 1;
                                        } else {
                                            if (isspace(calcBuffer[cursorPos - 1]) ||
                                                strchr(DELIMITERS, calcBuffer[cursorPos - 1])) break;
                                        }

                                        // move on loop
                                        cursorPos--;
                                    }
                                }
                            }

                        } else if (!strcmp(seq, ";5C")) {
                            // control+right, aka move right (word)
                            uint8_t rightWhite = 0;

                            if (cursorPos < len) {
                                if (strchr(DELIMITERS, calcBuffer[cursorPos])) {
                                    while (cursorPos < len && strchr(DELIMITERS, calcBuffer[cursorPos])) {
                                        cursorPos++;
                                    }

                                } else {
                                    while (cursorPos < len) {
                                        // skip trailling whitespace
                                        if (!rightWhite) {
                                            if (!isspace(calcBuffer[cursorPos])) rightWhite = 1;
                                        } else {
                                            if (isspace(calcBuffer[cursorPos]) ||
                                                strchr(DELIMITERS, calcBuffer[cursorPos])) break;
                                        }

                                        // move on loop
                                        cursorPos++;
                                    }
                                }
                            }

                        } else {
                            continue;
                        }

                        break;
                    }

                    default: continue; break;
                }
            }
        }

        if (c == 127 || c == 0x8) {
            // backspace
            if (cursorPos > 0) {
                // move everything back, accounting for cursorPos
                memmove(&calcBuffer[cursorPos - 1], &calcBuffer[cursorPos], len - cursorPos);
                cursorPos--;
                calcBuffer[--len] = '\0';
            }

        } else if (c == 0x17) {
            // ctrl + w, aka delete word
            uint8_t leftWhite = 0;

            if (cursorPos > 0) {
                if (strchr(DELIMITERS, calcBuffer[cursorPos - 1])) {
                    // if the cursor is already at a delimiter,
                    // delete only delimiters on loop
                    while (cursorPos > 0 && strchr(DELIMITERS, calcBuffer[cursorPos - 1])) {
                        // delete on loop
                        memmove(&calcBuffer[cursorPos - 1], &calcBuffer[cursorPos], len - cursorPos);
                        cursorPos--;
                        calcBuffer[--len] = '\0';
                    }

                } else {
                    while (cursorPos > 0) {
                        // skip trailling whitespace
                        if (!leftWhite) {
                            if (!isspace(calcBuffer[cursorPos - 1])) leftWhite = 1;
                        } else {
                            if (isspace(calcBuffer[cursorPos - 1]) ||
                                strchr(DELIMITERS, calcBuffer[cursorPos - 1])) break;
                        }

                        // delete on loop
                        memmove(&calcBuffer[cursorPos - 1], &calcBuffer[cursorPos], len - cursorPos);
                        cursorPos--;
                        calcBuffer[--len] = '\0';
                    }
                }
            }

        } else if (c == 0x18) {
            // ctrl + x, aka clear line
            len = 0;
            cursorPos = 0;
            memset(calcBuffer, '\0', sizeof(calcBuffer));

        } else if (c == 0xB) {
            // ctrl + k, aka clear everything right of cursor position
            if (calcBuffer[0] != '\0') {
                memset(calcBuffer + cursorPos, '\0', len - cursorPos);
                len = cursorPos;
            }

        } else if (c == 0x1) {
            // ctrl + a, aka move to start
            cursorPos = 0;

        } else if (c == 0x5) {
            // ctrl + e , aka move to end
            cursorPos = len;

        } else {
            // append (if character is valid, ofc)
            // we move the buffer to make space, like for when the cursor isnt at the end
            if (isprint(c)) {
                if (c == ',') c = '.';
                memmove(&calcBuffer[cursorPos + 1], &calcBuffer[cursorPos], len - cursorPos + 1);
                calcBuffer[cursorPos++] = c;
            }
            len++;
        }
    }

    putchar('\n');
    return retCode;
}
